File: slots.py
import re
from typing import TypedDict
from constants import Padding, parse_offsets, parse_anchors, parse_vector2, INDENT, i, format_float, format_vector2, fn
from rich import print
import traceback

registered_classes: list[type['Widget']] = []

class ParsedWidget(TypedDict): 
    props: dict
    children: list['ParsedWidget']

class Widget:
    Name: str
    DisplayName: str
    ClassName: str
    props: dict

    @property
    def var_name(self) -> str:
        return self.DisplayName[1:].split("_FONT")[0]

    def is_var(self) -> bool:
        return self.DisplayName.startswith("$")

    def __init_subclass__(cls) -> None:
        if(getattr(cls, "ClassName", None) is None):
            return
        
        registered_classes.append(cls)
    
    def __init__(self, object: ParsedWidget) -> None:
        self.Name = object['props'].get("Name", "").replace("\"", "")
        self.DisplayName = object['props'].get("DisplayLabel", "").replace("\"", "")

        if(self.DisplayName == ""):
            self.DisplayName = self.Name

        self.props = object['props']

    def codify(self, indent: int, parsed_objects: list['Widget']) -> str:
        return ""
    
    def __str__(self) -> str:
        return f"{self.__class__.__name__}(Name={self.Name})"

    def __repr__(self) -> str:
        return self.__str__()
    
# Slots
class Slot(Widget):
    Content: str
    
    variable: Widget | None = None
    widget: Widget | None = None

    def __init__(self, object: ParsedWidget) -> None:
        super().__init__(object)

        self.Content = object['props'].get("Content", "").split("'")[1].split("'")[0]

    def format_widget(self, indent: int, parsed_objects: list[Widget]) -> str:
        self.widget = next((obj for obj in parsed_objects if obj.Name == self.Content), None)

        if self.widget is None:
            return f"Script error: Widget {self.Content} not found\n"

        if self.widget.DisplayName.startswith("$external_"):
            name = self.widget.DisplayName[10:]
            return name + "\n\n"

        if self.widget.is_var():
            self.variable = self.widget
            
            # Trigger other variables
            self.widget.codify(indent, parsed_objects)

            return self.widget.var_name + "\n\n" # Put a newline at the end for better formatting

        return self.widget.codify(indent, parsed_objects) + "\n" # Also here
        
class CanvasSlot(Slot):
    ClassName: str = "/Script/UMG.CanvasPanelSlot"

    Anchors: tuple[float, float, float, float] | None
    Offsets: dict[str, float] | None
    Alignment: tuple[float, float] | None
    SizeToContent: bool = False
    ZOrder: int = 0
    
    def __init__(self, object: ParsedWidget) -> None:
        super().__init__(object)

        props = object['props']

        self.SizeToContent = props.get("bAutoSize", "False") == "True"
        
        self.Offsets = None
        self.Anchors = None
        self.Alignment = None

        layout_data = props.get("LayoutData", "")
        if(layout_data):
            layout_data = layout_data[1:-1]
            
            offsets = re.search(r"Offsets=\((.*?)\)", layout_data)
            anchors = re.search(r"Anchors=\((.*?)\)\)", layout_data)
            alignment = re.search(r"Alignment=\((.*?)\)", layout_data)

            if offsets:
                self.Offsets = parse_offsets(offsets.group(1))

                if(all(o == 0.0 for o in self.Offsets.values())):
                    self.Offsets = None

            if anchors:
                self.Anchors = parse_anchors(anchors.group(1))

            if alignment:
                self.Alignment = parse_vector2(alignment.group(1))

                if(self.Alignment == (0.0, 0.0)):
                    self.Alignment = None

        self.ZOrder = int(props.get("ZOrder", 0))

    def __str__(self) -> str:
        return f"CanvasSlot(Name={self.Name}, Anchors={self.Anchors}, Offsets={self.Offsets}, Alignment={self.Alignment}, SizeToContent={self.SizeToContent}, ZOrder={self.ZOrder}, Content={self.Content})\n"
    
    def codify(self, indent: int, parsed_objects: list[Widget]) -> str:
        result = f"{i(indent)}canvas_slot:\n"

        if self.ZOrder != 0:
            result += f"{i(indent + 1)}ZOrder := {self.ZOrder}\n"

        if self.Anchors:
            result += f"{i(indent + 1)}Anchors := Anchors({fn(self.Anchors[0])}, {fn(self.Anchors[1])}, {fn(self.Anchors[2])}, {fn(self.Anchors[3])})\n"

        result += f"{i(indent + 1)}SizeToContent := {'true' if self.SizeToContent else 'false'}\n"

        if self.Offsets:
            result += f"{i(indent + 1)}Offsets := Offsets({self.Offsets.get('Left', '0.0')}, {self.Offsets.get('Top', '0.0')}{f', {self.Offsets.get('Right', '100.0')}, {self.Offsets.get('Bottom', '30.0')}' if not self.SizeToContent else ''})\n"

        if self.Alignment:
            result += f"{i(indent + 1)}Alignment := {format_vector2(self.Alignment)}\n"

        return result + f"{i(indent + 1)}Widget := " + self.format_widget(indent + 1, parsed_objects)

class StackBoxSlot(Slot):
    ClassName: str = "/Script/UMG.StackBoxSlot"

    distribution: float
    padding: Padding
    horizontal_alignment: str
    vertical_alignment: str
    Content: str

    def __init__(self, object: ParsedWidget) -> None:
        super().__init__(object)

        props = object['props']

        self.vertical_alignment = props.get("VerticalAlignment", "VAlign_Fill").replace("VAlign_", "")
        self.horizontal_alignment = props.get("HorizontalAlignment", "HAlign_Fill").replace("HAlign_", "")

        self.padding = Padding.parse(props.get("Padding", "Padding(Left=0.0)"))
        self.distribution = -1

        size = props.get("Size", None)
        if(size):
            size_rule = re.search(r"SizeRule=([a-zA-Z]+)", size)

            if(size_rule and size_rule.group(1) == "Fill"):
                self.distribution = 1.0

                value = re.search(r"Value=([0-9\.-]+)", size)
                if(value):
                    self.distribution = float(value.group(1))

    def __str__(self) -> str:
        return f"StackBoxSlot(Name={self.Name}, Padding={self.padding}, HorizontalAlignment={self.horizontal_alignment}, VerticalAlignment={self.vertical_alignment}, Content={self.Content})\n"

    def codify(self, indent: int, parsed_objects: list[Widget]) -> str:
        result = f"{i(indent)}stack_box_slot:\n"

        if(not self.padding.is_empty()):
            result += self.padding.codify(indent + 1)
        
        result += f"{i(indent + 1)}HorizontalAlignment := horizontal_alignment.{self.horizontal_alignment}\n"
        result += f"{i(indent + 1)}VerticalAlignment := vertical_alignment.{self.vertical_alignment}\n"

        if self.distribution != -1:
            result += f"{i(indent + 1)}Distribution := {format_float(self.distribution)}.Maybe()\n"

        return result + f"{i(indent + 1)}Widget := " + self.format_widget(indent + 1, parsed_objects)
    
# OverlaySlot is exactly the same as StackBoxSlot, but with a different class name
class OverlaySlot(StackBoxSlot):
    ClassName: str = "/Script/UMG.OverlaySlot"
    
    def __str__(self) -> str:
        return f"OverlaySlot(Name={self.Name}, Padding={self.padding}, HorizontalAlignment={self.horizontal_alignment}, VerticalAlignment={self.vertical_alignment}, Content={self.Content})\n"

    def codify(self, indent: int, parsed_objects: list[Widget]) -> str:
        return super().codify(indent, parsed_objects).replace("stack_box_slot", "overlay_slot", 1)

# Slotables
class Slotable(Widget):
    slots: list[Slot]

    def __init__(self, object: ParsedWidget, slot_class: type[Slot]):
        super().__init__(object)

        self.slots = []

        for child in object['children']:
            try:
                if(slot_class.ClassName in child['props'].get('ExportPath', "") and child["props"].get("Content")):
                    slot = slot_class(child)

                    # Ignore slots with __ignore in the content
                    if("__ignore" in slot.Content):
                        continue

                    self.slots.append(slot)

            except Exception as e:
                print(f"Error parsing slot {child['props'].get("ExportPath")}")
                traceback.print_exc()
                pass

        # Sort slots
        slot_indexes: list[tuple[int, str]] = []
        for prop in object['props']:
            if(prop.startswith("Slots")):
                slot_index = int(prop.split("(")[1].split(")")[0])
                slot_indexes.append((
                    slot_index, 
                    object['props'][prop].replace("\"", "").split("'")[1].split("'")[0]
                ))

        slot_indexes.sort(key=lambda x: x[0])
        
        sorted_slots = []
        for index, slot_name in slot_indexes:
            s = next((slot for slot in self.slots if slot.Name == slot_name), None)
            if s:
                sorted_slots.append(s)

        self.slots = sorted_slots

    def __str__(self) -> str:
        return f"{self.__class__.__name__}(Name={self.Name}, Slots={self.slots})"
    
    def __repr__(self) -> str:
        return self.__str__()

    def format_slots(self, indent: int, parsed_objects: list[Widget]) -> str:
        if len(self.slots) == 0:
            return ""
        
        result = f"{i(indent - 1)}Slots := array:\n"
        
        for slot in self.slots:
            result += slot.codify(indent, parsed_objects)
        
        return result

class Canvas(Slotable):
    ClassName: str = "/Script/UMG.CanvasPanel"

    def __init__(self, object: ParsedWidget):
        super().__init__(object, CanvasSlot)

    def codify(self, indent: int, parsed_objects: list[Widget]) -> str:
        return f"canvas:\n" + self.format_slots(indent + 2, parsed_objects)

class StackBox(Slotable):
    ClassName: str = "/Script/UMG.StackBox"

    vertical: bool

    def __init__(self, object: ParsedWidget):
        super().__init__(object, StackBoxSlot)

        self.vertical = object['props'].get("Orientation", "Horizontal").replace("Orient_", "") == "Vertical"

    def codify(self, indent: int, parsed_objects: list[Widget]) -> str:
        result = \
f'''stack_box:
{i(indent + 1)}Orientation := orientation.{"Vertical" if self.vertical else "Horizontal"}
'''

        return result + self.format_slots(indent + 2, parsed_objects)
    
class Overlay(Slotable):
    ClassName: str = "/Script/UMG.Overlay"

    def __init__(self, object: ParsedWidget):
        super().__init__(object, OverlaySlot)

    def codify(self, indent: int, parsed_objects: list[Widget]) -> str:
        return f"overlay:\n" + self.format_slots(indent + 2, parsed_objects)

File: src/slots.ts
import { ParsedWidget } from "./types";
import {
    Padding,
    parseOffsets,
    parseAnchors,
    parseVector2,
    formatVector2,
    i,
    fn,
} from "./constants";
import type { UMGParser } from "./parser";

export const registeredClasses: Array<{
    ClassName: string;
    new (parser: UMGParser, object: ParsedWidget, ...args: unknown[]): Widget;
}> = [];

export class Widget {
    Name: string;
    DisplayName: string;
    ClassName?: string;
    SimpleName = "Widget";
    props: Record<string, string>;
    parser: UMGParser;

    get varName(): string {
        return this.DisplayName.substring(1); // .split("_FONT")[0]
    }

    isVar(): boolean {
        return this.DisplayName.startsWith("$");
    }

    constructor(parser: UMGParser, object: ParsedWidget) {
        this.parser = parser;
        this.Name = (object.props.Name ?? "").replace(/"/g, "");
        this.DisplayName = (object.props.DisplayLabel ?? "").replace(/"/g, "");
        if (this.DisplayName === "") this.DisplayName = this.Name;
        this.props = object.props;
    }

    codify(_indent: number, _parsedObjects: Widget[]): string {
        // Fallback to export path
        let widgetClass = this.props.Class ?? `Script error: Class for ${this.Name} not found`;

        // Remove project name from path
        const secondSlash = widgetClass.indexOf("/", 1);
        if (secondSlash !== -1) {
            widgetClass = widgetClass.substring(secondSlash + 1);

            // Remove part after last dot
            const lastDot = widgetClass.lastIndexOf(".");
            if (lastDot !== -1) {
                widgetClass = widgetClass.substring(0, lastDot);
            }
        }

        const widgetPath = widgetClass.replace(/\//g, ".") + "{}";
        return `${widgetPath}\n`;
    }
}

// Slots
export class Slot extends Widget {
    Content: string;
    variable: Widget | null = null;
    widget: Widget | null = null;

    constructor(parser: UMGParser, object: ParsedWidget) {
        super(parser, object);
        this.Content = (object.props.Content ?? "").split("'")[1]?.split("'")[0] ?? "";
    }

    formatWidget(indent: number, parsedObjects: Widget[]): string {
        this.widget = parsedObjects.find((obj) => obj.Name === this.Content) ?? null;

        if (this.widget === null) {
            return `Script error: Widget ${this.Content} not found\n`;
        }

        if (this.widget.DisplayName.startsWith("$external_")) {
            const name = this.widget.DisplayName.substring(10);
            return name + "\n\n";
        }

        if (this.widget.isVar()) {
            this.variable = this.widget;
            this.widget.codify(indent, parsedObjects);
            return this.widget.varName + "\n\n";
        }

        return this.widget.codify(indent, parsedObjects) + "\n";
    }
}

export class CanvasSlot extends Slot {
    static ClassName = "/Script/UMG.CanvasPanelSlot";
    override SimpleName = "CanvasSlot";

    Anchors: [number, number, number, number] | null;
    Offsets: Record<string, number> | null;
    Alignment: [number, number] | null;
    SizeToContent: boolean;
    ZOrder: number;

    constructor(parser: UMGParser, object: ParsedWidget) {
        super(parser, object);

        const props = object.props;
        this.SizeToContent = props.bAutoSize === "True";
        this.Offsets = null;
        this.Anchors = null;
        this.Alignment = null;

        let layoutData = props.LayoutData ?? "";
        if (layoutData) {
            layoutData = layoutData.substring(1, layoutData.length - 1);

            const offsets = layoutData.match(/Offsets=\((.*?)\)/);
            const anchors = layoutData.match(/Anchors=\((.*?)\)\)/);
            const alignment = layoutData.match(/Alignment=\((.*?)\)/);

            if (offsets) {
                this.Offsets = parseOffsets(offsets[1]);
                if (Object.values(this.Offsets).every((o) => o === 0.0)) {
                    this.Offsets = null;
                }
            }

            if (anchors) {
                this.Anchors = parseAnchors(anchors[1]);
            }

            if (alignment) {
                this.Alignment = parseVector2(alignment[1]);
                if (this.Alignment[0] === 0.0 && this.Alignment[1] === 0.0) {
                    this.Alignment = null;
                }
            }
        }

        this.ZOrder = parseInt(props.ZOrder ?? "0", 10);
    }

    override codify(indent: number, parsedObjects: Widget[]): string {
        let result = `${i(indent)}canvas_slot:\n`;

        if (this.ZOrder !== 0) {
            result += `${i(indent + 1)}ZOrder := ${this.ZOrder}\n`;
        }

        if (this.Anchors) {
            result += `${i(indent + 1)}Anchors := Anchors(${fn(this.Anchors[0])}, ${fn(this.Anchors[1])}, ${fn(this.Anchors[2])}, ${fn(this.Anchors[3])})\n`;
        }

        result += `${i(indent + 1)}SizeToContent := ${this.SizeToContent ? "true" : "false"}\n`;

        if (this.Offsets) {
            const left = this.Offsets.Left ?? 0.0;
            const top = this.Offsets.Top ?? 0.0;
            const sizeStr = this.SizeToContent
                ? ""
                : `, ${this.Offsets.Right ?? 100.0}, ${this.Offsets.Bottom ?? 30.0}`;
            result += `${i(indent + 1)}Offsets := Offsets(${left}, ${top}${sizeStr})\n`;
        }

        if (this.Alignment) {
            result += `${i(indent + 1)}Alignment := ${formatVector2(this.Alignment)}\n`;
        }

        return result + `${i(indent + 1)}Widget := ` + this.formatWidget(indent + 1, parsedObjects);
    }
}
registeredClasses.push(CanvasSlot as any);

export class StackBoxSlot extends Slot {
    static ClassName = "/Script/UMG.StackBoxSlot";
    override SimpleName = "StackBoxSlot";
    static DefaultVAlign = "VAlign_Fill";
    static DefaultHAlign = "HAlign_Fill";

    distribution: number;
    padding: Padding;
    horizontalAlignment: string;
    verticalAlignment: string;

    constructor(parser: UMGParser, object: ParsedWidget) {
        super(parser, object);

        const props = object.props;
        const ctor = this.constructor as typeof StackBoxSlot;

        this.verticalAlignment = (props.VerticalAlignment ?? ctor.DefaultVAlign).replace("VAlign_", "");
        this.horizontalAlignment = (props.HorizontalAlignment ?? ctor.DefaultHAlign).replace("HAlign_", "");
        this.padding = Padding.parse(props.Padding ?? "Padding(Left=0.0)");
        this.distribution = -1;

        const size = props.Size;
        if (size) {
            const sizeRule = size.match(/SizeRule=([a-zA-Z]+)/);
            if (sizeRule && sizeRule[1] === "Fill") {
                this.distribution = 1.0;
                const value = size.match(/Value=([0-9.\-]+)/);
                if (value) this.distribution = parseFloat(value[1]);
            }
        }
    }

    override codify(indent: number, parsedObjects: Widget[]): string {
        let result = `${i(indent)}stack_box_slot:\n`;

        if (!this.padding.isEmpty()) {
            result += this.padding.codify(indent + 1);
        }

        result += `${i(indent + 1)}HorizontalAlignment := horizontal_alignment.${this.horizontalAlignment}\n`;
        result += `${i(indent + 1)}VerticalAlignment := vertical_alignment.${this.verticalAlignment}\n`;

        if (this.distribution !== -1) {
            result += `${i(indent + 1)}Distribution := ${fn(this.distribution)}.Maybe()\n`;
        }

        return result + `${i(indent + 1)}Widget := ` + this.formatWidget(indent + 1, parsedObjects);
    }
}
registeredClasses.push(StackBoxSlot as any);

export class OverlaySlot extends StackBoxSlot {
    static override ClassName = "/Script/UMG.OverlaySlot";
    override SimpleName = "OverlaySlot";
    static override DefaultVAlign = "VAlign_Top";
    static override DefaultHAlign = "HAlign_Left";

    override codify(indent: number, parsedObjects: Widget[]): string {
        return super.codify(indent, parsedObjects).replace("stack_box_slot", "overlay_slot");
    }
}
registeredClasses.push(OverlaySlot as any);

// Slotables
export class Slotable extends Widget {
    slots: Slot[];

    constructor(
        parser: UMGParser,
        object: ParsedWidget,
        SlotClass: typeof CanvasSlot | typeof StackBoxSlot | typeof OverlaySlot,
    ) {
        super(parser, object);
        this.slots = [];

        for (const child of object.children) {
            try {
                if (
                    (child.props.ExportPath ?? "").includes(SlotClass.ClassName) &&
                    child.props.Content
                ) {
                    const slot = new SlotClass(parser, child);
                    if (slot.Content.includes("__ignore")) continue;
                    this.slots.push(slot);
                }
            } catch {
                // skip invalid slots
            }
        }

        // Sort slots by their index in the Slots(...) props
        const slotIndexes: Array<[number, string]> = [];
        for (const prop of Object.keys(object.props)) {
            if (prop.startsWith("Slots")) {
                const slotIndex = parseInt(prop.split("(")[1].split(")")[0], 10);
                const slotName = object.props[prop].replace(/"/g, "").split("'")[1]?.split("'")[0] ?? "";
                slotIndexes.push([slotIndex, slotName]);
            }
        }

        slotIndexes.sort((a, b) => a[0] - b[0]);

        const sortedSlots: Slot[] = [];
        for (const [, slotName] of slotIndexes) {
            const s = this.slots.find((slot) => slot.Name === slotName);
            if (s) sortedSlots.push(s);
        }

        this.slots = sortedSlots;
    }

    formatSlots(indent: number, parsedObjects: Widget[]): string {
        if (this.slots.length === 0) return "";

        let result = `${i(indent - 1)}Slots := array:\n`;
        for (const slot of this.slots) {
            result += slot.codify(indent, parsedObjects);
        }
        return result;
    }
}

export class Canvas extends Slotable {
    static ClassName = "/Script/UMG.CanvasPanel";
    override SimpleName = "Canvas";

    constructor(parser: UMGParser, object: ParsedWidget) {
        super(parser, object, CanvasSlot);
    }

    override codify(indent: number, parsedObjects: Widget[]): string {
        return "canvas:\n" + this.formatSlots(indent + 2, parsedObjects);
    }
}
registeredClasses.push(Canvas as any);

export class StackBox extends Slotable {
    static ClassName = "/Script/UMG.StackBox";
    override SimpleName = "StackBox";
    vertical: boolean;

    constructor(parser: UMGParser, object: ParsedWidget) {
        super(parser, object, StackBoxSlot);
        this.vertical = (object.props.Orientation ?? "Horizontal").replace("Orient_", "") === "Vertical";
    }

    override codify(indent: number, parsedObjects: Widget[]): string {
        let result = `stack_box:\n`;
        result += `${i(indent + 1)}Orientation := orientation.${this.vertical ? "Vertical" : "Horizontal"}\n`;
        return result + this.formatSlots(indent + 2, parsedObjects);
    }
}
registeredClasses.push(StackBox as any);

export class Overlay extends Slotable {
    static ClassName = "/Script/UMG.Overlay";
    override SimpleName = "Overlay";

    constructor(parser: UMGParser, object: ParsedWidget) {
        super(parser, object, OverlaySlot);
    }

    override codify(indent: number, parsedObjects: Widget[]): string {
        return "overlay:\n" + this.formatSlots(indent + 2, parsedObjects);
    }
}
registeredClasses.push(Overlay as any);
